packages/grafana-data/src/types/dataLink.ts

```diff
@@ -1,11 +1,32 @@
+import { ScopedVars } from './ScopedVars';
+
+/**
+ * Callback info for DataLink click events
+ */
+export interface DataLinkClickEvent<T = any> {
+  origin: T;
+  scopedVars: ScopedVars;
+  e?: any; // mouse/react event?
+}
+
 /**
  * Link configuration.  The values may contain variables that need to be
  * processed before running
  */
 export interface DataLink {
-  url: string;
   title: string;
   targetBlank?: boolean;
+
+  // 3: The URL if others did not set it first
+  url: string;
+
+  // 2: If exists, use this to construct the URL
+  // Not saved in JSON/DTO
+  onBuildHref?: (event: DataLinkClickEvent) => string;
+
+  // 1: If exists, handle click directly
+  // Not saved in JSON/DTO
```

Comment:
Yeah, but at some point we need to be able to serialize the handler right?  Here's some input on that https://stackoverflow.com/questions/7395686/how-can-i-serialize-a-function-in-javascript

Reply:
But after second thought, the onClick will most probably be some kind of behaviour specified on runtime (question is in what circumstances we define this behaviour and what will the behaviour be? i.e. change template variable value?). So maybe what will be needed to be stored in the DTO is some kind of a behaviour descriptor rather than the function itself. My thinking is maybe we have a registry of predefined behaviours that can be attached to data link?

Reply:
I think the main use is having the 'onclick' behavior set in the (frontend)  datasourco/panel. -- NOT the stored response

Reply:
I think we will also want predefined behaviors that are user configured and saved in the DTO and/or set by the datasource

Reply:
> I think the main use is having the 'onclick' behavior set in the (frontend) datasourco/panel. -- NOT the stored response

Yeah, of course, but we need to have a way to persist some of those behaviours somehow, I mean user specified ones? Or you mean that those behaviours will be some king of "internal" hooks for data links? Not sure what those hooks could be though, or what's on your mind.

Reply:
Eventually it would be something like 'open my modal'  or 'select this point' were another panel will show the details.  But all just brainstorming.